raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1791,6 +1993,17 @@ private Optional<Errors> validateVoterOnlyRequest(int remoteNodeId, int requestE
         }
     }
 
+    /**
+     * Return true if the voter key matches the local replica's key
+     */
+    private boolean isValidVoterKey(Optional<ReplicaKey> voterKey) {
+        return voterKey
+            .map(key ->
+                OptionalInt.of(key.id()).equals(nodeId) &&
+                key.directoryId().equals(Optional.of(nodeDirectoryId))
```

Comment:
Is it possible that the directoryID is null? No, because the voterKey only provided in newer version of request, right?

Reply:
`null`? No, since the default value for `Uuid` in Kafka data messages is `Uuid.ZERO_UUID`.

Having said that, since the new RPC adds both `voterId` and `voterDirectoryId`, if `voterId` is set (greater than or equal to 0), we assume that `voterDirectoryId` is set. If `voterId` is not set then `RaftUtil.voteRequestVoterKey` returns `Optional.empty()`.